### PR TITLE
Caching added similar to Swift

### PR DESCRIPTION
**Pull request by MaxHeimbrock**
## Background

Stacked cached token source in Rust, similar to swift: https://github.com/livekit/client-sdk-swift/blob/main/Sources/LiveKit/Token/CachingTokenSource.swift

### CODE REVIEW

File: livekit-token-source/src/caching.rs [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+        F: Fn(&TokenSourceFetchOptions, &TokenSourceResponse) -> bool + Send + Sync + 'static,
+    {
+        self.validator = Box::new(validator);
+        self
+    }
+
+    /// Removes the stored credentials, forcing the next fetch to hit the
+    /// underlying source.
+    pub async fn invalidate(&self) {
+        self.store.clear().await;
```
Comment (xianshijing-lk) [inline livekit-token-source/src/caching.rs:119]: what will happen if a fetch is ongoing here ?

File: livekit-token-source/src/caching.rs [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+        &self,
+        options: &TokenSourceFetchOptions,
+    ) -> TokenSourceResult<TokenSourceResponse> {
+        if let Some((cached_options, cached_response)) = self.store.retrieve().await {
+            if cached_options == *options && (self.validator)(&cached_options, &cached_response) {
+                return Ok(cached_response);
+            }
+        }
+
+        let response = self.source.fetch(options).await?;
```
Comment (xianshijing-lk) [inline livekit-token-source/src/caching.rs:141]: curiously, can multiple fetch() be called at the same time ? I wonder if we will need to protect such corner case.

File: livekit-token-source/tests/caching.rs [hunk 3]
```diff
@@ -0,0 +1,287 @@
...
+    ) -> (Self, Arc<AtomicUsize>, Arc<Mutex<Option<TokenSourceResponse>>>) {
+        let calls = Arc::new(AtomicUsize::new(0));
+        let next = Arc::new(Mutex::new(Some(response(token))));
+        (Self { calls: calls.clone(), next: next.clone() }, calls, next)
+    }
+}
+
+#[async_trait]
+impl TokenSourceConfigurable for CountingSource {
+    async fn fetch(
```
Comment (xianshijing-lk) [inline livekit-token-source/tests/caching.rs:80]: nit, should this fetch() has a .await to mimic the real use case ?

**Review by xianshijing-lk — APPROVED** [review]
some nits and questions, lgtm in general

File: livekit-token-source/src/caching.rs (hunk 1 above)
Comment (MaxHeimbrock) [reply livekit-token-source/src/caching.rs:119]: Added documentation:
```
/// A fetch already in flight is unaffected: it still resolves and stores
/// its response afterwards, repopulating the cache (last writer wins).
```

File: livekit-token-source/src/caching.rs (hunk 2 above)
Comment (MaxHeimbrock) [reply livekit-token-source/src/caching.rs:141]: Yes, and we don't protect against this, same as in Swift btw. This would be a bigger design. I think as other SDKs are not that advanced either yet, maybe we keep it in mind as a follow up.

File: livekit-token-source/tests/caching.rs (hunk 3 above)
Comment (MaxHeimbrock) [reply livekit-token-source/tests/caching.rs:80]: In these tests it would not change the behaviour, it would still run in sequence.

I added tests though for a token source that can mimic parallel in flight requests to test the existing contract of `last writer wins`.
